### PR TITLE
Add BioPrintingFacility promo card

### PR DESCRIPTION
**Pull request by vsrisuknimit**
https://www.fryxgames.se/produkt/spring-season-promo-2021/

image

image

image

### CODE REVIEW

File: src/cards/promo/BioPrintingFacility.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      return undefined;
+    });
+
+    if (availableAnimalCards.length === 1) {
+      const targetCard = availableAnimalCards[0];
+
+      return new OrOptions(
+        new SelectOption('Add 1 animal to ' + targetCard.name, 'Add animal', () => {
+          player.addResourceTo(targetCard, 1);
+          LogHelper.logAddResource(player, targetCard, 1);
```
Comment (nwai90) [inline src/cards/promo/BioPrintingFacility.ts:65]: Similarly `LogHelper.logAddResource(player, targetCard` suffices here

File: src/cards/promo/BioPrintingFacility.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    const availableAnimalCards = player.getResourceCards(ResourceType.ANIMAL);
+    player.energy -= 2;
+
+    if (availableAnimalCards.length === 0) {
+      player.plants += 2;
+      LogHelper.logGainStandardResource(player, Resources.PLANTS, 2);
+      return undefined;
+    }
+
+    const gainPlantOption = new SelectOption('Gain 2 plant', 'Gain plant', () => {
```
Comment (nwai90) [inline src/cards/promo/BioPrintingFacility.ts:53]: `Gain 2 plants` instead of `Gain 2 plant`?

File: src/CardName.ts [hunk 3]
```diff
@@ -462,6 +462,9 @@ export enum CardName {
     STANFORD_TORUS = 'Stanford Torus',
     SATURN_SURFING = 'Saturn Surfing',
 
+    // Promo cards
+    BIO_PRINTING_FACILITY = 'Bio Printing Facility',
```
Comment (nwai90) [inline src/CardName.ts:466]: For future: Group all promo cards together here. Does not need to be done in this PR

File: src/cards/promo/BioPrintingFacility.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      LogHelper.logGainStandardResource(player, Resources.PLANTS, 2);
+      return undefined;
+    });
+
+    if (availableAnimalCards.length === 1) {
+      const targetCard = availableAnimalCards[0];
+
+      return new OrOptions(
+        new SelectOption('Add 1 animal to ' + targetCard.name, 'Add animal', () => {
+          player.addResourceTo(targetCard, 1);
```
[the remark this answers is not in the document]
Comment (vsrisuknimit) [reply src/cards/promo/BioPrintingFacility.ts:64]: done

File: tests/cards/promo/BioPrintingFacility.spec.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+  it('Should act - single target', function() {
+    const smallanimals = new SmallAnimals();
+    player.playedCards.push(smallanimals);
+
+    const action = card.action(player);
+    expect(action instanceof OrOptions).is.true;
+    expect(action!.options).has.lengthOf(2);
+
+        action!.options[0].cb();
```
[the remark this answers is not in the document]
Comment (vsrisuknimit) [reply tests/cards/promo/BioPrintingFacility.spec.ts:45]: done

File: src/cards/promo/BioPrintingFacility.ts (hunk 2 above)
Comment (vsrisuknimit) [reply src/cards/promo/BioPrintingFacility.ts:53]: Gain plants as button label, done.

**Review by nwai90 — APPROVED** [review]

File: src/CardName.ts (hunk 3 above)
Comment (vsrisuknimit) [reply src/CardName.ts:466]: done. grouped all promo together.

File: src/cards/promo/BioPrintingFacility.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      cardType: CardType.ACTIVE,
+      name: CardName.BIO_PRINTING_FACILITY,
+      tags: [Tags.BUILDING],
+      cost: 7,
+
+      metadata: {
+        cardNumber: 'X36',
+        renderData: CardRenderer.builder((b) => {
+          b.action('Spend 2 energy to gain 2 plants OR to add 1 animal to ANOTHER card.', (eb) => {
+            eb.energy(2).digit.startAction.plants(2).or().animals(1).asterix();
```
Comment (chosta) [inline src/cards/promo/BioPrintingFacility.ts:28]: You can have 2 lines here just for better readability.
`eb.energy(2).digit.startAction.plants(2);
 eb.or().animals(1).asterix();
`

**Review by chosta — APPROVED** [review]

File: src/cards/promo/BioPrintingFacility.ts (hunk 6 above)
Comment (vsrisuknimit) [reply src/cards/promo/BioPrintingFacility.ts]: done

**Pull request merged**